server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2398,6 +2408,18 @@ public void activateWithPrimaryContext(final ReplicationTracker.PrimaryContext p
         assert getLocalCheckpoint() == primaryContext.getCheckpointStates().get(routingEntry().allocationId().getId())
             .getLocalCheckpoint() || indexSettings().getTranslogDurability() == Translog.Durability.ASYNC :
             "local checkpoint [" + getLocalCheckpoint() + "] does not match checkpoint from primary context [" + primaryContext + "]";
+        final Future<Void> future;
+        synchronized (mutex) {
+            future = relocationCondition;
+            relocationCondition = null;
+        }
+        if (future != null) {
+            try {
+                future.get();
```

Comment:
This is really lazy admittedly, I'd remove the blocking here for the final version to not waste a generic thread for a long period of time, but I figured it's much easier to quickly discuss the approach based on this than some tricky async impl.